File: src/midiscales.py
import math
from collections import OrderedDict
from settings import settings
from constants import NUM_PADS

midi_banks_chromatic = [
    [0 + i for i in range(16)],
    [4 + i for i in range(16)],
    [20 + i for i in range(16)],
    [36 + i for i in range(16)],
    [52 + i for i in range(16)],
    [68 + i for i in range(16)],
    [84 + i for i in range(16)],
    [100 + i for i in range(16)],
    [111 + i for i in range(16)]
]

scale_root_notes = [('C', 0),
                    ('Db', 1), 
                    ('D', 2), 
                    ('Eb', 3), 
                    ('E', 4), 
                    ('F', 5), 
                    ('Gb', 6), 
                    ('G', 7),
                    ('Ab', 8), 
                    ('A', 9), 
                    ('Bb', 10), 
                    ('B', 11)]


scale_intervals = OrderedDict({
    "maj": [2, 2, 1, 2, 2, 2, 1],
    "min": [2, 1, 2, 2, 1, 2, 2],
    "harm_min": [2, 1, 2, 2, 1, 3, 1],
    "mel_min": [2, 1, 2, 2, 2, 2, 1],
    "dorian": [2, 1, 2, 2, 2, 1, 2],
    "phrygian": [1, 2, 2, 2, 1, 2, 2],
    "lydian": [2, 2, 2, 1, 2, 2, 1]
})


def generate_midi_notes_in_scale(root, scale_intervals):
    """Generate MIDI notes in a scale, split into NUM_PADS-sized banks."""
    octave = 1 
    midi_notes = []
    cur_note = root
    
    # Add the root note first
    midi_notes.append(root)

    for scale_interval in scale_intervals:
        cur_note = cur_note + scale_interval
        midi_notes.append(cur_note)

    base_notes = midi_notes
    while cur_note < 127:
        for note in base_notes:
            cur_note = note + (12 * octave)
            if cur_note > 127:
                break
            midi_notes.append(cur_note)
        octave = octave + 1

    # Split into pad sets of size NUM_PADS
    midi_notes_pad_mapped = []
    num_banks = math.ceil(len(midi_notes) / NUM_PADS)  # Calculate minimum banks needed
    
    for i in range(num_banks):
        start_idx = i * NUM_PADS
        end_idx = start_idx + NUM_PADS
        padset = midi_notes[start_idx:end_idx]

        # Pad with last valid note if bank is not full
        if len(padset) < NUM_PADS:
            last_valid_note = padset[-1] if padset else midi_notes[-1]
            padset.extend([last_valid_note] * (NUM_PADS - len(padset)))

        midi_notes_pad_mapped.append(padset)

    return midi_notes_pad_mapped

# Store scale definitions instead of pre-generated scales for lazy loading
scale_definitions = [
    ("chromatic", None),  # Special case - no intervals needed
]

# Add the scale definitions from scale_intervals
for scale_name, interval in scale_intervals.items():
    scale_definitions.append((scale_name, interval))

def get_scale_notes(scale_idx, root_idx):
    if scale_idx == 0:  # Chromatic scale
        return midi_banks_chromatic
    
    # Get the scale definition
    scale_name, intervals = scale_definitions[scale_idx]
    root_note = scale_root_notes[root_idx][1]  # Get the MIDI note number
    return generate_midi_notes_in_scale(root_note, intervals)

def get_current_scale_notes():
    return get_scale_notes(settings.scale_idx, settings.rootnote_idx)

def get_scale_display_text():
    """Return multi-line display text for the current scale."""
    if settings.scale_idx == 0:  # special handling for chromatic
        disp_text = ["     Chromatic",
                     "",
                     f"        {settings.scale_idx+1}/{NUM_SCALES}"]
    else:
        scale_name = scale_definitions[settings.scale_idx][0]
        root_name = scale_root_notes[settings.rootnote_idx][0]
        disp_text = [f"     {root_name} {scale_name}",
                     "",
                     f"{settings.rootnote_idx+1}/{NUM_ROOTS}           {settings.scale_idx+1}/{NUM_SCALES}"]
    return disp_text

def get_midi_banks_chromatic():
    return midi_banks_chromatic

NUM_SCALES = len(scale_definitions)
NUM_ROOTS = len(scale_root_notes)

File: src/midiscales.cpp
#include "midiscales.h"
#include "settings.h"

namespace midiscales {

// midi_banks_chromatic starting offsets: 0,4,20,36,52,68,84,100,111 (+0..15 each)
static const uint8_t CHROMATIC_STARTS[9] = {0, 4, 20, 36, 52, 68, 84, 100, 111};

const char *const SCALE_NAMES[NUM_SCALES] = {
    "chromatic", "maj", "min", "harm_min", "mel_min", "dorian", "phrygian", "lydian"
};

const char *const ROOT_NAMES[NUM_ROOTS] = {
    "C", "Db", "D", "Eb", "E", "F", "Gb", "G", "Ab", "A", "Bb", "B"
};

// scale_intervals, same order as SCALE_NAMES[1..]
static const uint8_t SCALE_INTERVALS[NUM_SCALES - 1][7] = {
    {2, 2, 1, 2, 2, 2, 1}, // maj
    {2, 1, 2, 2, 1, 2, 2}, // min
    {2, 1, 2, 2, 1, 3, 1}, // harm_min
    {2, 1, 2, 2, 2, 2, 1}, // mel_min
    {2, 1, 2, 2, 2, 1, 2}, // dorian
    {1, 2, 2, 2, 1, 2, 2}, // phrygian
    {2, 2, 2, 1, 2, 2, 1}, // lydian
};

// generate_midi_notes_in_scale(), including the original's exact fill/pad behavior.
static uint8_t generate_banks(uint8_t root, const uint8_t intervals[7], ScaleBanks &out) {
    uint8_t midi_notes[160];
    int count = 0;
    int cur_note = root;

    midi_notes[count++] = root;
    for (int i = 0; i < 7; i++) {
        cur_note += intervals[i];
        midi_notes[count++] = (uint8_t)cur_note;
    }

    int base_count = count; // first-octave notes (root + 7 intervals)
    int octave = 1;
    while (cur_note < 127) {
        for (int i = 0; i < base_count; i++) {
            cur_note = midi_notes[i] + (12 * octave);
            if (cur_note > 127) {
                break;
            }
            midi_notes[count++] = (uint8_t)cur_note;
        }
        octave++;
    }

    // Split into banks of NUM_PADS, padding a short final bank with its last note
    uint8_t num_banks = (count + C::NUM_PADS - 1) / C::NUM_PADS;
    if (num_banks > MAX_BANKS) {
        num_banks = MAX_BANKS;
    }
    for (uint8_t b = 0; b < num_banks; b++) {
        int start = b * C::NUM_PADS;
        uint8_t last_valid = midi_notes[count - 1];
        for (uint8_t p = 0; p < C::NUM_PADS; p++) {
            int idx = start + p;
            if (idx < count) {
                out.notes[b][p] = midi_notes[idx];
                last_valid = midi_notes[idx];
            } else {
                out.notes[b][p] = last_valid;
            }
        }
    }
    out.numBanks = num_banks;
    return num_banks;
}

uint8_t get_scale_notes(uint8_t scale_idx, uint8_t root_idx, ScaleBanks &out) {
    if (scale_idx == 0) { // chromatic
        for (uint8_t b = 0; b < 9; b++) {
            for (uint8_t p = 0; p < C::NUM_PADS; p++) {
                out.notes[b][p] = CHROMATIC_STARTS[b] + p;
            }
        }
        out.numBanks = 9;
        return 9;
    }
    uint8_t root_note = root_idx; // scale_root_notes maps name -> 0..11
    return generate_banks(root_note, SCALE_INTERVALS[scale_idx - 1], out);
}

uint8_t get_current_scale_notes(ScaleBanks &out) {
    return get_scale_notes(settings.scale_idx, settings.rootnote_idx, out);
}

void get_scale_display_text(String out[3]) {
    if (settings.scale_idx == 0) { // special handling for chromatic
        out[0] = "     Chromatic";
        out[1] = "";
        out[2] = "        " + String(settings.scale_idx + 1) + "/" + String(NUM_SCALES);
    } else {
        out[0] = "     " + String(ROOT_NAMES[settings.rootnote_idx]) + " " + String(SCALE_NAMES[settings.scale_idx]);
        out[1] = "";
        out[2] = String(settings.rootnote_idx + 1) + "/" + String(NUM_ROOTS) + "           " +
                 String(settings.scale_idx + 1) + "/" + String(NUM_SCALES);
    }
}

} // namespace midiscales
